File: internal/modelmap/modelmap.go
// Package modelmap parses and resolves user-defined model substitutions.
//
// A substitution changes only the model ID sent to the Codex backend. Grok
// Build continues to address the source model ID, which makes it possible to
// override built-in entries such as grok-build or grok-4.5 without changing
// Grok Build's agent and subagent configuration.
package modelmap

import (
	"fmt"
	"sort"
	"strings"
	"unicode"

	"github.com/bengHak/grok-build-proxy/internal/catalog"
)

const fastSuffix = "-fast"

// Entry is one source-to-target model substitution as supplied by the user.
type Entry struct {
	Source string
	Target string
}

// Resolution is the effective upstream model for one requested model ID.
type Resolution struct {
	Requested string
	Model     string
	Fast      bool
	Mapped    bool
	Chain     []string
}

// EffectiveModelID returns the user-facing target ID, including the local
// -fast suffix when priority service is enabled. The actual upstream request
// sends Model without the suffix and sets service_tier separately.
func (r Resolution) EffectiveModelID() string {
	if r.Fast && r.Model != "" {
		return r.Model + fastSuffix
	}
	return r.Model
}

// Map is an immutable set of model substitutions. Its zero value is usable.
type Map struct {
	direct  map[string]string
	entries []Entry
}

// Parse accepts comma-, semicolon-, or newline-separated source=target pairs.
// Whitespace around pairs, sources, and targets is ignored.
func Parse(value string) (Map, error) {
	parts := strings.FieldsFunc(value, func(r rune) bool {
		return r == ',' || r == ';' || r == '\n' || r == '\r'
	})
	if len(parts) == 0 {
		return Map{}, nil
	}

	direct := make(map[string]string, len(parts))
	entries := make([]Entry, 0, len(parts))
	for _, part := range parts {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		source, target, ok := strings.Cut(part, "=")
		if !ok {
			return Map{}, fmt.Errorf("invalid model substitution %q: expected source=target", part)
		}
		source = strings.TrimSpace(source)
		target = strings.TrimSpace(target)
		if source == "" || target == "" {
			return Map{}, fmt.Errorf("invalid model substitution %q: source and target are required", part)
		}
		if err := validateModelID(source); err != nil {
			return Map{}, fmt.Errorf("invalid source model %q: %w", source, err)
		}
		if err := validateModelID(target); err != nil {
			return Map{}, fmt.Errorf("invalid target model %q: %w", target, err)
		}
		if strings.Contains(target, "=") {
			return Map{}, fmt.Errorf("invalid model substitution %q: too many '=' separators", part)
		}
		if source == target {
			return Map{}, fmt.Errorf("invalid model substitution %q: source and target are identical", part)
		}
		if previous, exists := direct[source]; exists {
			return Map{}, fmt.Errorf("duplicate model substitution for %q (%q and %q)", source, previous, target)
		}
		direct[source] = target
		entries = append(entries, Entry{Source: source, Target: target})
	}

	mapping := Map{direct: direct, entries: entries}
	for _, entry := range entries {
		if _, err := mapping.resolve(entry.Source); err != nil {
			return Map{}, err
		}
	}
	return mapping, nil
}

func validateModelID(value string) error {
	for _, r := range value {
		if unicode.IsSpace(r) {
			return fmt.Errorf("model IDs cannot contain whitespace")
		}
		if unicode.IsControl(r) {
			return fmt.Errorf("model IDs cannot contain control characters")
		}
	}
	return nil
}

// Empty reports whether no substitutions are configured.
func (m Map) Empty() bool {
	return len(m.direct) == 0
}

// Len returns the number of explicitly configured substitutions.
func (m Map) Len() int {
	return len(m.direct)
}

// Entries returns substitutions in the order supplied by the user.
func (m Map) Entries() []Entry {
	return append([]Entry(nil), m.entries...)
}

// SortedEntries returns substitutions sorted by source ID for deterministic
// diagnostics and API responses.
func (m Map) SortedEntries() []Entry {
	entries := m.Entries()
	sort.Slice(entries, func(i, j int) bool {
		return entries[i].Source < entries[j].Source
	})
	return entries
}

// Resolve returns the final upstream model. Substitutions may chain. A -fast
// suffix on the request or any target in the chain enables priority service and
// is removed before the final model ID is sent upstream.
func (m Map) Resolve(requested string) Resolution {
	resolution, err := m.resolve(requested)
	if err != nil {
		// Parse validates every reachable chain, and Map's fields are private, so
		// this is defensive only. Falling back to the normalized request is safer
		// than routing to an arbitrary partial target.
		base, fast := catalog.NormalizeID(requested)
		return Resolution{
			Requested: strings.TrimSpace(requested),
			Model:     base,
			Fast:      fast,
			Mapped:    false,
			Chain:     []string{strings.TrimSpace(requested)},
		}
	}
	return resolution
}

func (m Map) resolve(requested string) (Resolution, error) {
	requested = strings.TrimSpace(requested)
	baseRequested, requestedFast := catalog.NormalizeID(requested)
	if requested == "" || baseRequested == "" {
		return Resolution{Requested: requested}, nil
	}

	current := requested
	fast := requestedFast
	chain := []string{requested}
	visited := make(map[string]int, len(m.direct)+1)

	for {
		lookup := current
		next, found := m.direct[lookup]
		if !found {
			base, hasFast := catalog.NormalizeID(current)
			fast = fast || hasFast
			lookup = base
			next, found = m.direct[lookup]
			if !found {
				current = base
				break
			}
		} else {
			_, hasFast := catalog.NormalizeID(current)
			fast = fast || hasFast
		}

		if first, seen := visited[lookup]; seen {
			cycle := append(append([]string(nil), chain[first:]...), lookup)
			return Resolution{}, fmt.Errorf("model substitution cycle: %s", strings.Join(cycle, " -> "))
		}
		visited[lookup] = len(chain) - 1
		current = strings.TrimSpace(next)
		chain = append(chain, current)
	}

	finalModel, targetFast := catalog.NormalizeID(current)
	fast = fast || targetFast
	return Resolution{
		Requested: requested,
		Model:     finalModel,
		Fast:      fast,
		Mapped:    len(chain) > 1 || finalModel != baseRequested,
		Chain:     chain,
	}, nil
}

// String returns a stable comma-separated representation.
func (m Map) String() string {
	entries := m.SortedEntries()
	parts := make([]string, 0, len(entries))
	for _, entry := range entries {
		parts = append(parts, entry.Source+"="+entry.Target)
	}
	return strings.Join(parts, ",")
}

File: src/modelmap.rs
use crate::catalog::normalize_id;
use anyhow::{Result, bail};
use std::collections::{HashMap, HashSet};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Entry {
    pub source: String,
    pub target: String,
}
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Resolution {
    pub requested: String,
    pub model: String,
    pub fast: bool,
    pub mapped: bool,
    pub chain: Vec<String>,
}
impl Resolution {
    pub fn effective_model_id(&self) -> String {
        if self.fast && !self.model.is_empty() {
            format!("{}-fast", self.model)
        } else {
            self.model.clone()
        }
    }
}

#[derive(Clone, Debug, Default)]
pub struct ModelMap {
    direct: HashMap<String, String>,
    entries: Vec<Entry>,
}
impl ModelMap {
    pub fn parse(value: &str) -> Result<Self> {
        let mut direct = HashMap::new();
        let mut entries = Vec::new();
        for raw in value
            .split([',', ';', '\n', '\r'])
            .map(str::trim)
            .filter(|s| !s.is_empty())
        {
            let Some((source, target)) = raw.split_once('=') else {
                bail!("invalid model substitution {raw:?}: expected source=target")
            };
            let (source, target) = (source.trim(), target.trim());
            if source.is_empty() || target.is_empty() {
                bail!("invalid model substitution {raw:?}: source and target are required")
            }
            if target.contains('=') {
                bail!("invalid model substitution {raw:?}: too many '=' separators")
            }
            if source.chars().any(|c| c.is_whitespace() || c.is_control())
                || target.chars().any(|c| c.is_whitespace() || c.is_control())
            {
                bail!("model IDs cannot contain whitespace or control characters")
            }
            if source == target {
                bail!("invalid model substitution {raw:?}: source and target are identical")
            }
            if direct.insert(source.into(), target.into()).is_some() {
                bail!("duplicate model substitution for {source:?}")
            }
            entries.push(Entry {
                source: source.into(),
                target: target.into(),
            });
        }
        let map = Self { direct, entries };
        for e in &map.entries {
            map.try_resolve(&e.source)?;
        }
        Ok(map)
    }
    fn try_resolve(&self, requested: &str) -> Result<Resolution> {
        let requested = requested.trim().to_string();
        let (base_requested, mut fast) = normalize_id(&requested);
        if base_requested.is_empty() {
            return Ok(Resolution {
                requested,
                model: String::new(),
                fast: false,
                mapped: false,
                chain: vec![],
            });
        }
        let mut current = requested.clone();
        let mut chain = vec![requested.clone()];
        let mut visited = HashSet::new();
        loop {
            let (base, has_fast) = normalize_id(&current);
            fast |= has_fast;
            let (lookup, next) = if let Some(next) = self.direct.get(&current) {
                (current.clone(), Some(next))
            } else {
                (base.clone(), self.direct.get(&base))
            };
            let Some(next) = next else {
                current = base;
                break;
            };
            if !visited.insert(lookup.clone()) {
                chain.push(lookup);
                bail!("model substitution cycle: {}", chain.join(" -> "))
            }
            current = next.trim().into();
            chain.push(current.clone());
        }
        let (model, target_fast) = normalize_id(&current);
        fast |= target_fast;
        Ok(Resolution {
            requested,
            mapped: chain.len() > 1 || model != base_requested,
            model,
            fast,
            chain,
        })
    }
    pub fn resolve(&self, requested: &str) -> Resolution {
        self.try_resolve(requested).unwrap_or_else(|_| {
            let (model, fast) = normalize_id(requested);
            Resolution {
                requested: requested.trim().into(),
                model,
                fast,
                mapped: false,
                chain: vec![requested.trim().into()],
            }
        })
    }
    pub fn entries(&self) -> &[Entry] {
        &self.entries
    }
    pub fn len(&self) -> usize {
        self.direct.len()
    }
    pub fn is_empty(&self) -> bool {
        self.direct.is_empty()
    }
    pub fn stable_string(&self) -> String {
        let mut entries = self.entries.clone();
        entries.sort_by(|a, b| a.source.cmp(&b.source));
        entries
            .into_iter()
            .map(|e| format!("{}={}", e.source, e.target))
            .collect::<Vec<_>>()
            .join(",")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn chains_and_fast() {
        let m = ModelMap::parse("alias=middle,middle=gpt-5.6-sol-fast").unwrap();
        let r = m.resolve("alias");
        assert_eq!(r.model, "gpt-5.6-sol");
        assert!(r.fast && r.mapped);
    }
    #[test]
    fn rejects_cycles() {
        assert!(ModelMap::parse("a=b,b=a").is_err());
    }
}
